lib/llm/src/kv_router/publisher.rs

```diff
@@ -72,6 +87,373 @@ fn start_publish_task(
     });
 }
 
+// vLLM and SGLang use multi-processing to launch engine-core processes
+// We use zmq to publish events from these processes to a socket
+// For more info on zmq: https://zeromq.org/
+// This publisher reads those events and publishes them to NATS
+// The indexer will get the events from NATS and put them in the global prefix tree.
+pub struct KvEventPublisherFromZmq {
+    kv_block_size: usize,
+    processor_handle: Option<tokio::task::JoinHandle<()>>,
+    zmq_handle: Option<tokio::task::JoinHandle<()>>,
+    zmq_token: Option<dynamo_runtime::CancellationToken>,
+    warning_count: Arc<AtomicU32>,
+}
```

Comment:
There probably is a better way to consolidate `KvEventPublisherFromZmq` and `KvEventPublisher`, if you consider any kind of event publisher has an incoming edge and outgoing edge, both shares the same outgoing edge "pushing event message to nats". The difference is how they get the event from the first place, where `KvEventPublisher` receives events from active API calling and `zmq` from pulling the queue. Then you can have single event publisher abstraction and construct and set the "incoming edge" at runtime (also "outgoing edge" if needed to expand it in the future).